### ISSUE TITLE
[Feature request] Structured pruning support zeroing respective output channel biases

### ISSUE THREAD

**Issue by vballoli** [pytorch/pytorch#38598]
## 🚀 Feature

`LnStructured` method doesn't zero the biases of the respective output channel who's output is zero.  Example:
```
import torch
from torch.nn.utils import prune
c = torch.nn.Conv2d(3,6,3,bias=True)
c_ = prune.ln_structured(c, 'weight', 0.3, 1, 0) # L1 norm along the output channels 
```
Assuming channel 0 among the 6 channels gets pruned,
__Expected behaviour__: c.weight[0] = 0 and c.bias[0] = 0
__Current behaviour__: c.weight[0] = 0
## Motivation
Non-zero bias along a channel will have unintended effects to for the output activations. 

## Pitch
Support passing an additional function that can manipulate a module's attributes based on the indices obtained in `compute_mask` to increase flexibility. 

I did think of submitting a PR that hard codes the dimension and pruning the bias attribute of the module to fix the issue(specific to LnStructured only), but it felt like it wasn't the best practice, so I wanted to raise this issue with the author(s) first.
 
@mickypaganini

**Comment by he-y** [comment]
Weight and bias are considered seperately.
Take a look at [pruning toturial](https://pytorch.org/tutorials/intermediate/pruning_tutorial.html).

**Comment by vballoli** [comment]
Yeah but specifically in structured pruning, pruning the bias depends on what channels are being pruned, which is different from what's happening in the 'bias', since the elements there are pruned relative to the bias values and are independent of the weights or channels themselves.
@he-y

**Comment by he-y** [comment]
How about copy the mask of 'weights' to 'bias'?
@vballoli

**Comment by vballoli** [comment]
That's certainly is a solution, but imo a better approach would be to add a hooks like functionality for a more general solution that adds flexibility instead of restricting and handling it case by case.

**Comment by mickypaganini** [comment]
It's intended as is. We cannot assume that everybody who prunes the weight parameter in a structured way will want to prune the bias as well. In any case, if that's the behavior you need, it should be easy enough to implement: use the weight mask to compute the bias mask and apply it using the [`custom_from_mask`](https://pytorch.org/docs/master/generated/torch.nn.utils.prune.custom_from_mask.html) pruning function.

**Comment by mickypaganini** [comment]
An example:
```python
import torch
from torch.nn.utils import prune
c = torch.nn.Conv2d(3,6,3,bias=True)
prune.ln_structured(c, 'weight', 0.3, 1, 0) # L1 norm along the output channels

# Now compute the corresponding bias mask by checking which channels sum to zero in the weight and zeroing those out in the bias as well 
bias_mask = (
   torch.sum(c.weight_mask, axis=(1, 2, 3)) != 0
).to(torch.float32)

prune.custom_from_mask(c, 'bias', bias_mask)
```

Note, btw, that this is only relevant when you use `prune.ln_structured` along `dim=0`, i.e. along the output direction.

**Comment by he-y** [comment]
@mickypaganini 
Do you think it's necessary to add some state-of-the-art pruning algorithm?

**Comment by mickypaganini** [comment]
@he-y 
If I'm understanding your question correctly, I do _not_ personally think that it's worth capturing those two extra lines of code I showed in the example into its own separate pruning technique, also because, as is, this would go against the logic of these pruning techniques acting on a tensor at a time. But if there is evidence that my solution above cannot work in certain cases, then I'm happy to revisit this.

**Comment by he-y** [comment]
@mickypaganini 
Sorry for my unclear question. Your example could solve the problem properly. I agree that it's **not** necessary to add these two lines into code.

My question is about some other pruning algorithms. Recently, lots of new pruning algorithms are proposed, but the [current implementation](https://github.com/pytorch/pytorch/blob/4fef3763dd7266195e8ef20c0a5d4dd1219afeb0/torch/nn/utils/prune.py) mainly contains norm-based pruning methods. So do you think it's necessary to add some state-of-the-art pruning methods?

Personally, adding more SOTA method could benefit the community. My expectation of pruning methods is similar to the [Classification model in Pytorch](https://pytorch.org/docs/stable/torchvision/models.html). In this way, different pruning methods could be included and compared.

Maybe I should open another issue for this.

**Comment by mickypaganini** [comment]
Oh! I see what you meant. Yes, I agree with you that the current module can certainly be expanded to support further mainstream pruning techniques beyond the magnitude-based one we currently provide, and I support the idea. Best to open a separate issue for that, so we can discuss what could be included, provide evidence from the literature, and get feedback from other interested users.
